node/neighbor.go

```diff
@@ -129,3 +130,88 @@ func (localNode *LocalNode) startConnectingToRandomNeighbors() {
+	if !ok {
+		panic("Overlay is not chord")
+	}
+
+	allNeighbors := localNode.GetNeighbors(filter)
+	chordNeighborIDs := make(map[string]struct{}, len(allNeighbors))
+
+	for _, succ := range c.Successors() {
+		chordNeighborIDs[string(succ.Id)] = struct{}{}
+	}
+
+	for _, pred := range c.Predecessors() {
+		chordNeighborIDs[string(pred.Id)] = struct{}{}
+	}
+
+	for _, fingers := range c.FingerTable() {
+		for _, finger := range fingers {
+			chordNeighborIDs[string(finger.Id)] = struct{}{}
+		}
+	}
+
+	chordNeighbors := make([]*RemoteNode, 0, len(allNeighbors))
+	randomNeighbors := make([]*RemoteNode, 0, len(allNeighbors))
+
+	for _, neighbor := range allNeighbors {
+		if neighbor.Id == nil {
+			continue
+		}
+
+		if _, ok := chordNeighborIDs[string(neighbor.Id)]; ok {
+			chordNeighbors = append(chordNeighbors, neighbor)
+			continue
+		}
+
+		if fingerIdx, _ := c.FingerTableIdxInRemoteNode(neighbor.Id); fingerIdx >= 0 {
+			chordNeighbors = append(chordNeighbors, neighbor)
+			continue
+		}
+
+		randomNeighbors = append(randomNeighbors, neighbor)
+	}
+
+	return chordNeighbors, randomNeighbors
+}
+
+func (localNode *LocalNode) GetGossipNeighbors(filter func(*RemoteNode) bool) []*RemoteNode {
+	var sampledNeighbors []*RemoteNode
+	chordNeighbors, randomNeighbors := localNode.splitNeighbors(filter)
+
+	numChordSamples := int(config.GossipSampleChordNeighbor * float32(len(chordNeighbors)))
+	if numChordSamples > 0 {
+		rand.Shuffle(len(chordNeighbors), func(i, j int) { chordNeighbors[i], chordNeighbors[j] = chordNeighbors[j], chordNeighbors[i] })
+		sampledNeighbors = append(sampledNeighbors, chordNeighbors[:numChordSamples]...)
+	}
+
+	numRandomSamples := int(config.GossipSampleRandomNeighbor * float32(len(randomNeighbors)))
+	if numRandomSamples > 0 {
+		rand.Shuffle(len(randomNeighbors), func(i, j int) { randomNeighbors[i], randomNeighbors[j] = randomNeighbors[j], randomNeighbors[i] })
+		sampledNeighbors = append(sampledNeighbors, randomNeighbors[:numRandomSamples]...)
+	}
+
+	return sampledNeighbors
+}
+
+func (localNode *LocalNode) GetVotingNeighbors(filter func(*RemoteNode) bool) []*RemoteNode {
+	var sampledNeighbors []*RemoteNode
+	chordNeighbors, randomNeighbors := localNode.splitNeighbors(filter)
+
+	numChordSamples := int(config.VotingSampleChordNeighbor * float32(len(chordNeighbors)))
+	if numChordSamples > 0 {
+		rand.Shuffle(len(chordNeighbors), func(i, j int) { chordNeighbors[i], chordNeighbors[j] = chordNeighbors[j], chordNeighbors[i] })
+		sampledNeighbors = append(sampledNeighbors, chordNeighbors[:numChordSamples]...)
+	}
+
+	numRandomSamples := int(config.VotingSampleRandomNeighbor * float32(len(randomNeighbors)))
+	if numRandomSamples > 0 {
+		rand.Shuffle(len(randomNeighbors), func(i, j int) { randomNeighbors[i], randomNeighbors[j] = randomNeighbors[j], randomNeighbors[i] })
+		sampledNeighbors = append(sampledNeighbors, randomNeighbors[:numRandomSamples]...)
+	}
+
+	return sampledNeighbors
+}
```

Comment:
How about extract an internal function getNeighbors(filter, chordNum, randomNum) and invoke it as below:
```
GetGossipNeighbors(filter) {
    getNeighbors(filter, config.GossipSampleChordNeighbor, config.GossipSampleRandomNeighbor)
}
GetVotingNeighbors(filter) {
    getNeighbors(filter, config.VotingSampleChordNeighbor, config.VotingSampleRandomNeighbor)
}
```

Reply:
Good idea. Pushed the new version.